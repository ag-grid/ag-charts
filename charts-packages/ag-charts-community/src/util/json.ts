type LiteralProperties = 'shape' | 'data';
type SkippableProperties = 'axes' | 'series' | 'container' | 'customChartThemes';
type IsLiteralProperty<T, K extends keyof T> = K extends LiteralProperties ? true :
    T[K] extends Array<infer E> ? true :
    false;
type IsSkippableProperty<T, K extends keyof T> = K extends SkippableProperties ? true : false;

// Needs to be recursive when we move to TS 4.x+; only supports a maximum level of nesting right now.
export type DeepPartial<T> = {
    [P1 in keyof T]?:
        IsSkippableProperty<T, P1> extends true ? any :
        IsLiteralProperty<T, P1> extends true ? T[P1] :
        { [P2 in keyof T[P1]]?:
            IsSkippableProperty<T[P1], P2> extends true ? any :
            IsLiteralProperty<T[P1], P2> extends true ? T[P1][P2] :
            { [P3 in keyof T[P1][P2]]?:
                IsSkippableProperty<T[P1][P2], P3> extends true ? any :
                IsLiteralProperty<T[P1][P2], P3> extends true ? T[P1][P2][P3] :
                { [P4 in keyof T[P1][P2][P3]]?:
                    IsSkippableProperty<T[P1][P2][P3], P4> extends true ? any :
                    IsLiteralProperty<T[P1][P2][P3], P4> extends true ? T[P1][P2][P3][P4] :
                    Partial<T[P1][P2][P3][P4]>
                }
            }
        }
};

/**
 * Performs a JSON-diff between a source and target JSON structure.
 * 
 * On a per property basis, takes the target property value where:
 * - types are different.
 * - type is primitive.
 * - type is array and length or content have changed.
 * 
 * Recurses for object types.
 * 
 * @param source starting point for diff
 * @param target target for diff vs. source
 * @param opts.stringify properties to stringify for comparison purposes
 * 
 * @returns `null` if no differences, or an object with the subset of properties that have changed.
 */
export function jsonDiff<T extends any>(source: T, target: T, opts?: { stringify: string[] }): Partial<T> | null {
    const { stringify = [] } = opts || {};
    const sourceType = classify(source);
    const targetType = classify(target);

    if (targetType === 'array') {
        if (sourceType !== 'array' || source.length !== target.length) {
            return [ ...(target as any) ];
        }

        if (target.some((targetElement: any, i: number) => jsonDiff(source?.[i], targetElement) != null)) {
            return [ ...(target as any) ];
        }

        return null;
    }

    const lhs = source || {} as any;
    const rhs = target || {} as any;

    const allProps = new Set([
        ...Object.keys(lhs),
        ...Object.keys(rhs),
    ]);

    let propsChangedCount = 0;
    const result: any = {};
    for (const prop of allProps) {
        // Cheap-and-easy equality check.
        if (lhs[prop] === rhs[prop]) { continue; }

        const take = (v: any) => {
            result[prop] = v;
            propsChangedCount++;
        };

        if (stringify.includes(prop)) {
            if (JSON.stringify(lhs[prop] !== JSON.stringify(rhs[prop]))) { 
                take(rhs[prop]);
            }
            continue;
        }

        const lhsType = classify(lhs[prop]);
        const rhsType = classify(rhs[prop]);
        if (lhsType !== rhsType) {
            // Types changed, just take RHS.
            take(rhs[prop]);
            continue;
        }

        if (rhsType === 'primitive' || rhsType === null) {
            take(rhs[prop]);
            continue;
        }

        if (rhsType === 'array' && lhs[prop].length !== rhs[prop].length) {
            // Arrays are different sizes, so just take target array.
            take(rhs[prop]);
            continue;
        }

        if (rhsType === 'class-instance') {
            // Don't try to do anything tricky with array diffs!
            take(rhs[prop]);
            continue;
        }

        if (rhsType === 'function' && lhs[prop] !== rhs[prop]) {
            take(rhs[prop]);
            continue;
        }

        const diff = jsonDiff(lhs[prop], rhs[prop], { stringify });
        if (diff !== null) {
            take(diff);
        }
    }

    return propsChangedCount === 0 ? null : result;
}

/**
 * Special value used by `jsonMerge` to signal that a property should be removed from the merged
 * output.
 */
export const DELETE = Symbol('<delete-property>') as any;
const NOT_SPECIFIED = Symbol('<unspecified-property>') as any;

/**
 * Merge together the provide JSON object structures, with the precedence of application running
 * from higher indexes to lower indexes.
 * 
 * Deep-clones all objects to avoid mutation of the inputs changing the output object. For arrays,
 * just performs a deep-clone of the entire array, no merging of elements attempted.
 * 
 * @param json all json objects to merge
 * 
 * @returns the combination of all of the json inputs
 */
export function jsonMerge<T>(...json: T[]): T {
    const jsonTypes = json.map(v => classify(v));
    if (jsonTypes.some(v => v === 'array')) {
        // Clone final array.
        const finalValue = json[json.length - 1];
        if (finalValue instanceof Array) {
            return finalValue.map((v) => {
                const type = classify(v);
                return type === 'array' ? jsonMerge([], v) : 
                    type === 'object' ? jsonMerge({}, v) :
                    v;
            }) as any;
        }

        return finalValue;
    }

    const result: any = {};
    const props = new Set(
        json.map((v) => v != null ? Object.keys(v) : [])
            .reduce((r, n) => r.concat(n), []),
    );

    for (const nextProp of props) {
        const values = json.map(j => j != null && nextProp in j ? (j as any)[nextProp] : NOT_SPECIFIED)
            .filter(v => v !== NOT_SPECIFIED);
        
        if (values.length === 0) {
            continue;
        }
        const lastValue = values[values.length - 1];
        if (lastValue === DELETE) {
            continue;
        }

        const types = values.map(v => classify(v));
        const type = types[0];
        if (types.some(t => t !== type && t !== null)) {
            // Short-circuit if mismatching types.
            result[nextProp] = lastValue;
            continue;
        }

        if (type === 'array' || type === 'object') {
            result[nextProp] = jsonMerge(...values);
        } else {
            // Just directly assign/overwrite.
            result[nextProp] = lastValue;
        }
    }

    return result;
}

/**
 * Recursively apply a JSON object into a class-hierarchy, optionally instantiating certain classes
 * by property name.
 * 
 * @param target to apply source JSON properties into
 * @param source to be applied
 * @param params.path path for logging/error purposes, to aid with pinpointing problems
 * @param params.matcherPath path for pattern matching, to lookup allowedTypes override.
 * @param params.skip property names to skip from the source
 * @param params.constructors dictionary of property name to class constructors for properties that
 *                            require object construction
 * @param params.allowedTypes overrides by path for allowed property types
 */
export function jsonApply<
    Target,
    Source extends DeepPartial<Target>,
>(
    target: Target,
    source?: Source,
    params: {
        path?: string,
        matcherPath?: string,
        skip?: (keyof Source | keyof Target)[],
        constructors?: Record<string, new () => any>,
        allowedTypes?: Record<string, ReturnType<typeof classify>[]>,
    } = {},
): Target {
    const {
        path = undefined,
        matcherPath = path ? path.replace(/(\[[0-9+]{1,}\])/i, '[]') : undefined,
        skip = [],
        constructors = {},
        allowedTypes = {},
    } = params;

    if (target == null) { throw new Error(`AG Charts - target is uninitialised: ${path || '<root>'}`); }
    if (source == null) { return target; }

    const targetType = classify(target);
    for (const property in source) {
        if (skip.indexOf(property) >= 0) { continue; }

        const newValue = source[property];
        const propertyPath = `${path ? path + '.' : ''}${property}`;
        const propertyMatcherPath = `${matcherPath ? matcherPath + '.' : ''}${property}`;
        const targetAny = (target as any);
        const targetClass = targetAny.constructor;
        const currentValue = targetAny[property];
        const ctr = constructors[property];
        try {
            const currentValueType = classify(currentValue);
            const newValueType = classify(newValue);

            if (targetType === 'class-instance' && !(property in target || targetAny.hasOwnProperty(property))) {
                console.warn(`AG Charts - unable to set [${propertyPath}] in ${targetClass?.name} - property is unknown`);
                continue;
            }

            const allowableTypes = allowedTypes[propertyMatcherPath] || [currentValueType];
            if (currentValueType === 'class-instance' && newValueType === 'object') {
                // Allowed, this is the common case! - do not error.
            } else if (currentValueType != null && newValueType != null && !allowableTypes.includes(newValueType)) {
                console.warn(`AG Charts - unable to set [${propertyPath}] in ${targetClass?.name} - can't apply type of [${newValueType}], allowed types are: [${allowableTypes}]`);
                continue;
            }

            if (newValueType === 'array') {
                targetAny[property] = newValue;
            } else if (newValueType === 'class-instance') {
                targetAny[property] = newValue;
            } else if (newValueType === 'object') {
                if (currentValue != null) {
                    jsonApply(currentValue, newValue as any, {...params, path: propertyPath, matcherPath: propertyMatcherPath });
                } else if (ctr != null) {
                    targetAny[property] = jsonApply(new ctr(), newValue as any, {...params, path: propertyPath, matcherPath: propertyMatcherPath });
                } else {
                    targetAny[property] = newValue;
                }
            } else {
                targetAny[property] = newValue;
            }
        } catch (error) {
            console.warn(`AG Charts - unable to set [${propertyPath}] in [${targetClass?.name}]; nested error is: ${error.message}`);
            continue;
        }
    }

    return target;
}

/**
 * Walk the given JSON object graphs, invoking the visit() callback for every object encountered.
 * Arrays are descended into without a callback, however their elements will have the visit()
 * callback invoked if they are objects.
 * 
 * @param json to traverse
 * @param visit callback for each non-primitive and non-array object found
 * @param opts.skip property names to skip when walking
 * @param jsons to traverse in parallel
 */
export function jsonWalk(
    json: any,
    visit: (classification: Classification, node: any, ...otherNodes: any[]) => void,
    opts: {
        skip?: string[],
    },
    ...jsons: any[]
) {
    const jsonType = classify(json);
    const skip = opts.skip || [];

    if (jsonType === 'array') {
        json.forEach((element: any, index: number) => {
            jsonWalk(element, visit, opts, ...jsons?.map(o => o?.[index]));
        });
        return;
    } else if (jsonType !== 'object') {
        return;
    }

    visit(jsonType, json, ...jsons);
    for (const property in json) {
        if (skip.indexOf(property) >= 0) { continue; }

        const value = json[property];
        const otherValues = jsons?.map(o => o?.[property]);
        const valueType = classify(value);

        if (valueType === 'object' || valueType === 'array') {
            jsonWalk(value, visit, opts, ...otherValues);
        }
    }
}

type Classification = 'array' | 'object' | 'primitive';
/**
 * Classify the type of a value to assist with handling for merge purposes.
 */
export function classify(value: any): 'array' | 'object' | 'function' | 'primitive' | 'class-instance' | null {
    if (value == null) {
        return null;
    } else if (value instanceof HTMLElement) {
        return 'primitive';
    } else if (value instanceof Array) {
        return 'array';
    } else if (value instanceof Date) {
        return 'primitive';
    } else if (typeof value === 'object' && value.constructor === Object) {
        return 'object';
    } else if (typeof value === 'function') {
        return 'function';
    } else if (typeof value === 'object' && value.constructor != null) {
        return 'class-instance';
    }

    return 'primitive';
}
