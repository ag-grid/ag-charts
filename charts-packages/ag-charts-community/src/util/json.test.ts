import { describe, expect, it, jest } from '@jest/globals';
import { jsonDiff, jsonMerge, DELETE, jsonWalk, jsonApply } from './json';

const FIXED_DATE = new Date(2022, 0, 27, 0, 0, 0, 0);

class TestApply {
    num?: number = undefined;
    str?: string = undefined;
    date?: Date = undefined;
    array?: number[] = undefined;
    recurse?: TestApply = undefined;

    constructor(params: {[K in keyof TestApply]?: TestApply[K]} = {}) {
        Object.assign(this, params);
    }
}

describe('json module', () => {
    describe('#jsonDiff', () => {
        describe('for trivial cases', () => {
            it('should return null for no diff', () => {
                const cases: {}[] = [
                    {},
                    { a: 1, b: { c: 'abc', d: () => 'test' } },
                    { a: [{ foo: 'bar' }], b: { c: 'abc', d: () => 'test' } },
                ];

                for (const testCase of cases) {
                    expect(jsonDiff(testCase, testCase)).toBeNull();
                }

                expect(jsonDiff({}, {})).toBeNull();
            });

            it('should correctly diff primitives', () => {
                const source = {};
                const target = {
                    foo: 'bar',
                    hello: 123,
                    alice: FIXED_DATE,
                    func: (test) => test,
                };

                const diff = jsonDiff(source, target);
                expect(diff).toMatchSnapshot();
            });
        });

        describe('for non-trivial cases', () => {
            it('should correctly diff complex object structures', () => {
                const source = {
                    foo: { bar1: 1 },
                    hello1: { nested: { nestedX2: { primitive: 'abc' } } },
                    unchanging: { readonly: 1 },
                    changing: 'abc',
                    removed: 123,
                    removed2: { nested: { nestedX2: { primitive: 'abc' } } },
                };
                const target = {
                    foo: { bar1: 2 },
                    hello1: {
                        nested: { nestedX2: { primitive: 'abc', added: 123 } },
                        nestedAdd: { primitive: 123 },
                    },
                    unchanging: { readonly: 1 },
                    changing: '123',
                };

                const diff = jsonDiff(source, target as any);
                expect(diff).toMatchSnapshot();
                expect(diff).toHaveProperty('foo.bar1', target.foo.bar1);
                expect(diff).toHaveProperty('hello1.nested.nestedX2.added', 123);
                expect(diff).toHaveProperty('changing', target.changing);
                expect(diff).toHaveProperty('removed', undefined);
                expect(diff).toHaveProperty('removed2', undefined);
                expect(diff).not.toHaveProperty('unchanging');
                expect(diff).not.toHaveProperty('hello1.nestedX2.primitive');
            });

            it('should correctly diff identical arrays', () => {
                const source = {
                    foo: [1, 2, 3 , 4],
                };

                const diff = jsonDiff(source, source);
                expect(diff).toBeNull();
            });

            it('should correctly diff function changes in arrays', () => {
                const source = {
                    foo: [{ fn: () => 'hello-world!'}],
                };
                const target = {
                    foo: [{ fn: () => 'foo-bar!?!?!'}],
                };

                const diff = jsonDiff(source, target as any);
                expect(diff).toMatchSnapshot();
                expect(diff).toHaveProperty(['foo', '0', 'fn'], target.foo[0].fn);
            });
        });
    });

    describe('#jsonMerge', () => {
        describe('for trivial cases', () => {
            it('should merge primitives correctly', () => {
                const fns = [() => 'call-me1', () => 'call-me2', () => 'call-me3'];
                const base = {
                    no: 1,
                    foo: 'bar',
                    fn: fns[0],
                    date: FIXED_DATE,
                };
                const mergee1 = {
                    no: 2,
                    foo2: 'bar2',
                    fn2: fns[1],
                    date2: FIXED_DATE,
                };
                const mergee2 = {
                    no2: 2,
                    foo3: 'bar2',
                    fn2: fns[2],
                    date3: FIXED_DATE,
                };

                const merge = jsonMerge(base, mergee1 as any, mergee2);
                expect(merge).toMatchSnapshot();
                expect(merge).toHaveProperty('no', mergee1.no);
                expect(merge).toHaveProperty('no2', mergee2.no2);
                expect(merge).toHaveProperty('foo', base.foo);
                expect(merge).toHaveProperty('foo2', mergee1.foo2);
                expect(merge).toHaveProperty('foo3', mergee2.foo3);
                expect(merge).toHaveProperty('fn', base.fn);
                expect(merge).toHaveProperty('fn2', mergee2.fn2);
                expect(merge).toHaveProperty('date', base.date);
                expect(merge).toHaveProperty('date2', mergee1.date2);
                expect(merge).toHaveProperty('date3', mergee2.date3);
            });

            it('should merge array properties correctly', () => {
                const base: any = { a: [[{ x: 1 }, { y: 1 }], [{ m: 2, n: 2 }]], b: [1, 2, 3, 4, 5, 6] };
                const mergee1: any = { a: [], b: [] };
                const mergee2: any = { a: [[{ x2: 1 }, { y2: 1 }], [{ m2: 2, n2: 2 }]], c: [10, 9, 8, 7, 6] };

                const merge = jsonMerge(base, mergee1, mergee2);
                expect(merge).toMatchSnapshot();
                expect(merge).toHaveProperty('a', mergee2.a);
                expect(merge).toHaveProperty('b', mergee1.b);
                expect(merge).toHaveProperty('c', mergee2.c);
            });

            it('should merge arrays correctly', () => {
                const base: any = [[{ x: 1 }, { y: 1 }], [{ m: 2, n: 2 }]];
                const mergee1: any = [];
                const mergee2: any = [[{ x2: 1 }, { y2: 1 }], [{ m2: 2, n2: 2 }]];

                const merge = jsonMerge(base, mergee1, mergee2);
                expect(merge).toMatchSnapshot();
                expect(merge).toEqual(mergee2);
            });

            it('should honour DELETE value', () => {
                const base = { a: 1, b: 'b', c: FIXED_DATE, d: { 1: 1, 2: 2, 3: 3 } };
                const mergee1 = {};
                const mergee2 = { c: DELETE, d: { 2: DELETE } };

                const merge = jsonMerge(base, mergee1, mergee2);
                expect(merge).toMatchSnapshot();
                expect(merge).not.toHaveProperty('c');
                expect(merge).not.toHaveProperty('d.2');
            });

            it('should take highest precedent value when types conflict', () => {
                const base = { a: [[{ x: 1 }, { y: 1 }], [{ m: 2, n: 2 }]], b: [1, 2, 3, 4, 5, 6] };
                const mergee1 = { a: {}, b: {} };
                const mergee2 = { a: 'a' };

                const merge = jsonMerge(base, mergee1, mergee2 as any);
                expect(merge).toMatchSnapshot();
                expect(merge).toHaveProperty('a', mergee2.a);
                expect(merge).toHaveProperty('b', mergee1.b);
            });
        });
    });

    describe('#jsonWalk', () => {
        it('should visit no nodes for no object', () => {
            for(const test of [undefined, null, 'a', 1, FIXED_DATE]) {
                const cb = jest.fn();
                jsonWalk(test, cb, {}, test);
                expect(cb).toHaveBeenCalledTimes(0);
            }
        });

        it('should not visit property nodes for no object', () => {
            for(const test of [undefined, null, 'a', 1, FIXED_DATE]) {
                const wrappedTest = { test };

                const cb = jest.fn();
                jsonWalk(wrappedTest, cb, {}, wrappedTest);
                expect(cb).toHaveBeenCalledWith('object', wrappedTest, wrappedTest);
                expect(cb).toHaveBeenCalledTimes(1);
            }
        });

        it('should only visit one node for a trivial object', () => {
            const walked1 = { a: 1, b: 2, c: 'c', d: FIXED_DATE };
            const walked2 = { a: 2, b: 3, c: 'd', d: FIXED_DATE };

            const cb = jest.fn();
            jsonWalk(walked1, cb, {}, walked2);
            expect(cb).toHaveBeenCalledTimes(1);
            expect(cb).toHaveBeenCalledWith('object', walked1, walked2);
        });

        it('should visit every node for a non-trivial object', () => {
            const walked1 = {
                a: 1,
                b: 2,
                c: 'c',
                d: FIXED_DATE,
                child1: { foo: 'bar' },
                child2: { hello: 'world', child3: { x: 'x' } },
            };
            const walked2 = { a: 2, b: 3, c: 'd', d: FIXED_DATE, child1: { foo: 'bar' } };

            const cb = jest.fn();
            jsonWalk(walked1, cb, {}, walked2);
            expect(cb).toHaveBeenCalledWith('object', walked1, walked2);
            expect(cb).toHaveBeenCalledWith('object', walked1.child1, walked2.child1);
            expect(cb).toHaveBeenCalledWith('object', walked1.child2, undefined);
            expect(cb).toHaveBeenCalledWith('object', walked1.child2.child3, undefined);
            expect(cb).toHaveBeenCalledTimes(4);
        });

        it('should visit every node of an array', () => {
            const walked1 = [{a:1}, {b:2}, {c:3}, {d:4}];
            const walked2 = [{x:1}, {y:2}, {z:3}];

            const cb = jest.fn();
            jsonWalk(walked1, cb, {}, walked2);
            expect(cb).toHaveBeenCalledWith('object', walked1[0], walked2[0]);
            expect(cb).toHaveBeenCalledWith('object', walked1[1], walked2[1]);
            expect(cb).toHaveBeenCalledWith('object', walked1[2], walked2[2]);
            expect(cb).toHaveBeenCalledWith('object', walked1[3], undefined);
            expect(cb).toHaveBeenCalledTimes(4);
        });

        it('should visit every node of an array property', () => {
            const walked1 = { prop1: [{a:1}, {b:2}, {c:3}, {d:4}] };
            const walked2 = { prop1: [{x:1}, {y:2}, {z:3}] };

            const cb = jest.fn();
            jsonWalk(walked1, cb, {}, walked2);
            expect(cb).toHaveBeenCalledWith('object', walked1, walked2);
            expect(cb).toHaveBeenCalledWith('object', walked1.prop1[0], walked2.prop1[0]);
            expect(cb).toHaveBeenCalledWith('object', walked1.prop1[1], walked2.prop1[1]);
            expect(cb).toHaveBeenCalledWith('object', walked1.prop1[2], walked2.prop1[2]);
            expect(cb).toHaveBeenCalledWith('object', walked1.prop1[3], undefined);
            expect(cb).toHaveBeenCalledTimes(5);
        });

        it('should skip specified properties', () => {
            const walked1 = {
                a: 1,
                b: 2,
                c: 'c',
                d: FIXED_DATE,
                child1: { foo: 'bar' },
                child2: { hello: 'world', child3: { x: 'x' } },
            };
            const walked2 = { a: 2, b: 3, c: 'd', d: FIXED_DATE, child1: { foo: 'bar' } };

            const cb = jest.fn();
            jsonWalk(walked1, cb, { skip: ['child1', 'child3'] }, walked2);
            expect(cb).toHaveBeenCalledWith('object', walked1, walked2);
            expect(cb).toHaveBeenCalledWith('object', walked1.child2, undefined);
            expect(cb).toHaveBeenCalledTimes(2);
        });
    });

    describe('#jsonApply', () => {
        const json = {
            str: 'test-string', num: 123, date: FIXED_DATE, array: [1,2,3,4],
            recurse: { str: 'test-string2', num: 789, date: FIXED_DATE, array: [1,2,3,4] }
        };

        it('should be able to populate an existing object graph', () => {
            const target = new TestApply({ recurse: new TestApply() });
            jsonApply(target, json);
            expect(target.str).toEqual(json.str);
            expect(target.num).toEqual(json.num);
            expect(target.date).toEqual(json.date);
            expect(target.array).toEqual(json.array);
            expect(target.recurse).toBeInstanceOf(TestApply);
            expect(target.recurse.str).toEqual(json.recurse.str);
            expect(target.recurse.num).toEqual(json.recurse.num);
            expect(target.recurse.date).toEqual(json.recurse.date);
            expect(target.recurse.array).toEqual(json.recurse.array);
        });

        it('should be able to instantiate a new object graph', () => {
            const target = new TestApply();
            jsonApply(target, json, { constructors: { recurse: TestApply }});
            expect(target.str).toEqual(json.str);
            expect(target.num).toEqual(json.num);
            expect(target.date).toEqual(json.date);
            expect(target.array).toEqual(json.array);
            expect(target.recurse).toBeInstanceOf(TestApply);
            expect(target.recurse.str).toEqual(json.recurse.str);
            expect(target.recurse.num).toEqual(json.recurse.num);
            expect(target.recurse.date).toEqual(json.recurse.date);
            expect(target.recurse.array).toEqual(json.recurse.array);
        });

        it('should skip specified properties', () => {
            const target = new TestApply();
            jsonApply(target, json, { skip: ['str'], constructors: { recurse: TestApply }});
            expect(target.str).toEqual(undefined);
            expect(target.recurse.str).toEqual(undefined);
        });

        it('should error on unrecognised properties', () => {
            const badJson = { foo: 'bar' };
            const target = new TestApply();

            console.warn = jest.fn();
            jsonApply(target, badJson as any);
            expect(console.warn).toBeCalledWith("AG Charts - unable to set [foo] in TestApply - property is unknown");
        });

        it('should error on incompatible properties', () => {
            const badJson = { recurse: 'foo' };
            const target = new TestApply({ recurse: new TestApply() });

            console.warn = jest.fn();
            jsonApply(target, badJson as any);
            expect(console.warn).toBeCalledWith("AG Charts - unable to set [recurse] in TestApply - can't apply type of [primitive], allowed types are: [class-instance]");
        });

        it('should allow application of property type overrides', () => {
            const target = new TestApply({ recurse: new TestApply({ str: 'string'} )});
            const json = { recurse: { str: () => 'test' } };
            const opts = {
                path: 'series[0]',
                allowedTypes: {'series[].recurse.str': ['function' as const]},
                constructors: { recurse: TestApply },
            };

            jsonApply(target, json as any, opts);
            expect(target.recurse.str).toEqual(json.recurse.str);
        });

    });
});
