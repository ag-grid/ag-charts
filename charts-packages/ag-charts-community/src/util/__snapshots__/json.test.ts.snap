// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json module #jsonDiff for non-trivial cases should correctly diff complex object structures 1`] = `
Object {
  "changing": "123",
  "foo": Object {
    "bar1": 2,
  },
  "hello1": Object {
    "nested": Object {
      "nestedX2": Object {
        "added": 123,
      },
    },
    "nestedAdd": Object {
      "primitive": 123,
    },
  },
  "removed": undefined,
  "removed2": undefined,
}
`;

exports[`json module #jsonDiff for non-trivial cases should correctly diff function changes in arrays 1`] = `
Object {
  "foo": Array [
    Object {
      "fn": [Function],
    },
  ],
}
`;

exports[`json module #jsonDiff for trivial cases should correctly diff primitives 1`] = `
Object {
  "alice": 2022-01-27T00:00:00.000Z,
  "foo": "bar",
  "func": [Function],
  "hello": 123,
}
`;

exports[`json module #jsonMerge for trivial cases should honour DELETE value 1`] = `
Object {
  "a": 1,
  "b": "b",
  "d": Object {
    "1": 1,
    "3": 3,
  },
}
`;

exports[`json module #jsonMerge for trivial cases should merge array properties correctly 1`] = `
Object {
  "a": Array [
    Array [
      Object {
        "x2": 1,
      },
      Object {
        "y2": 1,
      },
    ],
    Array [
      Object {
        "m2": 2,
        "n2": 2,
      },
    ],
  ],
  "b": Array [],
  "c": Array [
    10,
    9,
    8,
    7,
    6,
  ],
}
`;

exports[`json module #jsonMerge for trivial cases should merge arrays correctly 1`] = `
Array [
  Array [
    Object {
      "x2": 1,
    },
    Object {
      "y2": 1,
    },
  ],
  Array [
    Object {
      "m2": 2,
      "n2": 2,
    },
  ],
]
`;

exports[`json module #jsonMerge for trivial cases should merge primitives correctly 1`] = `
Object {
  "date": 2022-01-27T00:00:00.000Z,
  "date2": 2022-01-27T00:00:00.000Z,
  "date3": 2022-01-27T00:00:00.000Z,
  "fn": [Function],
  "fn2": [Function],
  "foo": "bar",
  "foo2": "bar2",
  "foo3": "bar2",
  "no": 2,
  "no2": 2,
}
`;

exports[`json module #jsonMerge for trivial cases should take highest precedent value when types conflict 1`] = `
Object {
  "a": "a",
  "b": Object {},
}
`;
